query/common/hll.go

```diff
@@ -814,3 +845,262 @@ var hllBiases = []float64{
+	r.buffer.Write(bs[:])
+	strErr := err.Error()
+	padding := (8 - (len(strErr) & 7)) & 8
+	r.buffer.Write([]byte(strErr))
+	if padding > 0 {
+		paddingBytes := make([]byte, padding)
+		r.buffer.Write(paddingBytes)
+	}
+}
+
+// GetBytes returns the underlying bytes.
+func (r *HLLQueryResults) GetBytes() []byte {
+	return r.buffer.Bytes()
+}
+
+// BuildVectorsFromHLLResult traverses input HLL query result and builds byte slices
+// result must have HLL in the leave nodes
+// this function is useful when converting HLL query result to it's binary format
+// dimDataTypes stores types of each dimension, in the the same order as in query
+// dimensionVectorIndex stores re-ordered dimension index, sorted by dim datatype width
+func BuildVectorsFromHLLResult(result AQLQueryResult, dimDataTypes []memCom.DataType, enumDicts map[int]map[string]int, dimensionVectorIndex []int) (hllVector, dimVector, countVector []byte, err error) {
+	hllVector = []byte{}
+	countVector = []byte{}
+
+	dimVectors := make([][]byte, len(dimDataTypes))
+	validityVectors := make([][]byte, len(dimDataTypes))
+
+	err = traverseRecursive(0, map[string]interface{}(result), dimDataTypes, enumDicts, &hllVector, &countVector, dimVectors, validityVectors)
+	if err != nil {
+		return
+	}
+
+	dimVector = []byte{}
+	// append in descending order of width
+	for _, idx := range dimensionVectorIndex {
+		dimVector = append(dimVector, dimVectors[idx]...)
+	}
+	for _, idx := range dimensionVectorIndex {
+		dimVector = append(dimVector, validityVectors[idx]...)
+	}
+
+	return
+}
+
+// helper function that traverses HLL query result
+func traverseRecursive(level int, curr interface{}, dimDataTypes []memCom.DataType, enumDicts map[int]map[string]int, hllVector, countVector *[]byte, dimVectors, validityVectors [][]byte) (err error) {
+	switch curr.(type) {
+	case map[string]interface{}:
+		dimDataType := dimDataTypes[level]
+		dimValueBytes := memCom.DataTypeBytes(dimDataType)
+		for k, v := range curr.(map[string]interface{}) {
+			// convert dimension
+			var (
+				dataVal memCom.DataValue
+				isValid uint8 = 1
+			)
+			if NullString == k {
+				isValid = 0
+			} else {
+				if enumDict, ok := enumDicts[level]; ok {
+					enumVal := enumDict[k]
+					if dimValueBytes == 1 {
+						ui8 := uint8(enumVal)
+						dataVal = memCom.DataValue{
+							Valid:    true,
+							OtherVal: unsafe.Pointer(&ui8),
+						}
+					} else if dimValueBytes == 2 {
+						ui16 := uint16(enumVal)
+						dataVal = memCom.DataValue{
+							Valid:    true,
+							OtherVal: unsafe.Pointer(&ui16),
+						}
+					} else {
+						err = utils.StackError(nil, "data width %d doesn't match any enum", dimValueBytes)
+					}
+				} else {
+					dataVal, err = memCom.ValueFromString(k, dimDataType)
+					if err != nil {
+						return
+					}
+				}
+			}
+
+			if dataVal.Valid {
+				byteVal := dataVal.OtherVal
+				switch dimValueBytes {
+				case 8:
+					dimVectors[level] = append(dimVectors[level], (*[8]byte)(byteVal)[:]...)
```

Comment:
actually we need to append the dim value multiple times depending on the next level result size

Reply:
yep, fixed..